### DISCUSSION TITLE
How do i re-init a datepicker after a submit of a form?

### DISCUSSION THREAD

**Discussion by wschwarte** [livewire/livewire#8552] [1. Help] [answered]
I am new to livewire 3 and am trying to build a form. On the form is a text input with a datepicker (pickadate.js).  I init this datepicker in a document ready jQuery event. This is working. If i click the field i get the date picker and all is good.

The datepicker injects some html to the dom on init. 

Now when I submit the form, using 'wire:submit' and livewire reloads the form with the validation errors, the datepicker is stopped working, because the injected code is now gone. 

I think i need to call the init of the datepicker again, but am not sure how to do this. i tried putting it in 'Livewire:hool('morph.udated')', but that is not working and i am not even sure i would need this.

**Comment by joshhanley** [comment]
@wschwarte Use `wire:ignore` to stop livewire wiping out the dom changes from your library.

**Reply by wschwarte** [reply]
Where would i put that?  If i put it on the form-group div, the picker keeps working, but the validation stops working. 

This div gets injected below the input, and in addition some attributes are added to the input itself.

`<div class="picker" id="ophaaldatum_root" aria-hidden="true">...</div`

```
  <div class="form-group" >
    <input wire:model="ophaaldatum" type="text" class="form-control" placeholder="Gewenste ophaaldatum *" id="ophaaldatum" >
    
     <p class="help-block text-danger"></p>
     @error('ophaaldatum')
        <div class="help-block text-danger"><ul role="alert"><li>{{ $message }}</li></ul></div>
     @enderror
    </div>
```

**Comment by wschwarte** [comment]
Still lost here... 


I tried this in my component blade file, and used a `->dispatch('initCalendarPicker')`. 

in the console i can see 'event' being logged, also after form submit.
After form submit, the value of the textinput is 'testing'. 

But the pickadate() isn't doing anything. any ideas?

```
@script

<script>
 $wire.on('initCalendarPicker', () => {
     console.log('event');  

     $('#ophaaldatum').val('testing');

     $('#ophaaldatum').pickadate({
         today: '',
         disable: [6, 7]
     });
 });
</script>
@endscript
```

**Reply by PhiloNL** [reply]
See the wire:ignore docs for more information:
https://livewire.laravel.com/docs/wire-ignore

The easiest is to inspect the element using your browser console to see where the date picker is injected but you will likely end up with something like:

```
<div wire:ignore>
    <input wire:model="ophaaldatum" type="text" class="form-control" placeholder="Gewenste ophaaldatum *" id="ophaaldatum" >
</div>
```

This will stop Livewire from replacing this section of the HTML.

**Reply by wschwarte** [reply]
Hi, as you can see in the above post, i have already tried wire:ignore, but then when the form is submitted, the validation errors (if any) aren't working.

**Reply by PhiloNL** [reply]
I can't see wire:ignore in your post above but in that case you need to ensure that the validation messages (`@error('ophaaldatum')`) is outside the `wire:ignore` scope, for example:

```
  <div class="form-group" >
     <div wire:ignore>
    <input wire:model="ophaaldatum" type="text" class="form-control" placeholder="Gewenste ophaaldatum *" id="ophaaldatum" >
     </div>
    
     <p class="help-block text-danger"></p>
     @error('ophaaldatum')
        <div class="help-block text-danger"><ul role="alert"><li>{{ $message }}</li></ul></div>
     @enderror
</div>
```

The errors won't appear if they are inside the `wire:ignore` scope as Livewire will ignore any changes for that section.

**Reply by PhiloNL** [reply]
Also, see the following working example, in this case, wire:ignore doesn't seem to be necessary:
https://wirebox.app/b/4qp61

**Reply by wschwarte** [reply]
Hi, so i tried these things. The extra div for the wire:model seems to work so the picker keeps popping up. What i realized now is this datepicker isn't 'compatible' with how livewire works.

The date is not sent with wire:submit, so when i wire:model the ophaaldatum field, it is always empty in the backend class. 

Not sure how this works, or why it isn't working though. 

it it this datepicker: 
`https://amsul.ca/pickadate.js/`

```
    <link href="{{ asset('/js/pickadate.js-3.5.6/lib/themes/default.css') }}" rel="stylesheet"/>
    <link href="{{ asset('/js/pickadate.js-3.5.6/lib/themes/default.date.css') }}" rel="stylesheet"/>
```

```
<script type="text/javascript" src="https://code.jquery.com/jquery-1.12.4.min.js"></script>
<script type="text/javascript" src="https://cdn.usebootstrap.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
<script type="text/javascript" src="{{ asset('/js/pickadate.js-3.5.6/lib/compressed/picker.js') }}"></script>
<script type="text/javascript" src="{{ asset('/js/pickadate.js-3.5.6/lib/compressed/picker.date.js') }}"></script>
```

```
// in document.ready 

 $('#ophaaldatum').pickadate({
        today: '',
        clear: 'Reset',
        close: 'Sluiten',
        labelMonthNext: 'Next month',
        labelMonthPrev: 'Previous month',
        labelMonthSelect: 'Select a month',
        labelYearSelect: 'Select a year',

        formatSubmit: 'yyyy-mm-dd',
        format: 'dd-mm-yyyy',

        min: +3,
        firstDay: 1,
        disable: [6, 7]
    });
    ```

**Reply by PhiloNL** [reply, marked as answer]
You should be able to use the `on` function to listen for the date to be set and use this event to set the public property `date` using the Livewire helper (`@this.set('<public-property>', 'value');`):

```
picker.on('set', (date) => {
@this.set('date', date);
})
```

**Reply by wschwarte** [reply]
yes, that seemed to work. thanks. learned a lot of laravel/livewire from this simple form :-D 

```
    $input = $('#ophaaldatum').pickadate({
        today: '',
        clear: 'Reset',
        close: 'Sluiten',
        labelMonthNext: 'Next month',
        labelMonthPrev: 'Previous month',
        labelMonthSelect: 'Select a month',
        labelYearSelect: 'Select a year',

        formatSubmit: 'yyyy-mm-dd',
        format: 'dd-mm-yyyy',

        min: +3,
        firstDay: 1,
        disable: [6, 7]
    });

    picker = $input.pickadate('picker');

    picker.on('set', (date) => {
        const dateSelected = new Date(date.select);
        const readableDate = dateSelected.toLocaleDateString('nl-NL', {month: '2-digit', year: 'numeric', day: '2-digit'});

        @this.set('ophaaldatum', readableDate);
    })
```
